### ISSUE TITLE
Enhancement: Time to move away from libbass to add native support for M1-based Macs

### ISSUE THREAD

**Issue by mpogue2** [mpogue2/SquareDesk#606]
It's a PITA to continue to use libbass for music playback.  It's not well-supported -- it hasn't been updated in years.  It's also X86 only.  With the new M1 Macs, we should ideally go Native, with something that is more actively supported.

The main uses of libbass are:
- decoding (MP3/etc)
- playback (play/pause/resume/etc)
- sound effects
- pitch/tempo changing (independently)
- BPM detection
- EQ

Well, it turns out that we have more options now than we had 5 years ago when SquareDesk started.

- Qt6.2 QtMultimedia supports cross-platform music playback, and with some work, modifying audio on the way thru.
     Also supports sound effects (although they are WAV only right now, so we'd have to decode them into memory,
     or one-time convert and cache them).
- Rubber band library (seems to be good quality pitch/tempo changing = open source)
- MiniBPM (estimates BPM = open source)
- KFR (framework for optimized biquad filters for EQ, etc = open source)

Sadly, Superpowered, which supports all this stuff in a single cross-platform package, is no longer Open Source, so it's not a viable option.

So, this ticket is to get rid of libbass and libbassfx entirely, and replace them with code that is actively being developed and maintained, AND which also supports the M1 processor.  This might also help us to do iOS and Android versions at some point.

**Comment by mpogue2** [comment]
OK, as of 7d9096453270c7931dc76722b7c3c62cac1c80ec, I have a lot of the QtMultimedia stuff working (right now, it's on the M1 Mac ONLY -- Intel Mac, Windows, and Linux should NOT be affected at all).  I left in a lot of qDebug() messages, since this is still under active development, and because normal Intel Mac users won't see any of this code.

**What is working as of this commit:**
- load a song when you double click in songTable
- play/pause/stop a song
- seekBar tracks playback within the song
- clicking on seekBar jumps to that point in the song
- songPosition and totalSongLength values are correct
- stops at the end of a song correctly
- >>| and |<< buttons to go to next song and previous song work
- Playlist > Continuous Play works
- Autostart playback works

**What does NOT work yet (requires additional coding):**
- loops (the loop indicators are there, but the player does not actually loop yet)
- mono mode (mix down from stereo), requires DSP code
- mix slider (L/R), requires DSP code
- EQ (B/M/T) sliders, requires KFR and DSP code 
- Tempo/pitch sliders, requires RubberBand and DSP code
- VU Meter, requires DSP code
- soundEffects, requires QtMultimedia's QSoundEffects
- automatic BPM calculation at load time, requires MiniBPM code

**Comment by mpogue2** [comment]
Initial standalone versions of miniBPM and RubberBand are both in the tree now.  They are not hooked up to anything, but I verified that they compile on M1 Mac.

**Comment by mpogue2** [comment]
OK, I spent a bunch of time on trying to check "kfr" into git, and failed.  It seems that it already is somehow a "submodule", and getting it out of that mode is tricky.  If somebody wants to try getting kfr into the tree, that works for me.  Also, it's linked somehow to github's kfr, so we can't push there (and we don't want to either!).

In the mean time, here's how to get kfr and compile it to static libs:

```
cd <SquareDesk-DEV folder>
# clone kfr into the kfr folder, which should be there in git
git clone https://github.com/kfrlib/kfr.git
cd kfr
mkdir build
cd build
# make sure you downloaded CMake app into /Applications (we'll use the cmd line version inside that)
# make the makefiles, as per https://github.com/kfrlib/kfr
/Applications/CMake.app/Contents/bin/cmake -G"Unix Makefiles" -DENABLE_TESTS=OFF -DCMAKE_BUILD_TYPE=Release ..
# don't do the multiprocess -j, it didn't work for me
make
```
Now should have two .a files in there: libkfr_dft.a and libkfr_io.a.
Includes are in /Users/mpogue/clean3/SquareDesk/SquareDesk-DEV/kfr/include
Need to do:

Add this to command line:
`-Ipath_to_kfr/include
`
And this for libs, if needed
`-lkfr_dft -lkfr_io
`
C++17 mode must be enabled
`-std=c++17
`or
`-std=gnu++17`

**Comment by mpogue2** [comment]
References for implementing "DSP" mode:
How to use QAudioDecoder to read in a WAV file, then spit out to a QOutputDevice
https://stackoverflow.com/questions/41197576/how-to-play-mp3-file-using-qaudiooutput-and-qaudiodecoder

QAudioSink docs
https://doc-snapshots.qt.io/qt6-dev/qaudiosink.html

**Comment by mpogue2** [comment]
OK, here's where I am right now.
The QMediaPlayer seems to work fine.  Very reliable.  But, I can't get access to the data to do all this other stuff.

What we need for "DSP" mode is QAudioDecoder, and it is very flaky right now.  
- It frequently doesn't call bufferReady at all.  Often it won't call it unless finished() is connected.
- It stops in the middle of decoding a file, and just hangs.  isDecoding() is still true at this point, but no more bytes get passed to bufferReady.  Before that, it looks like everything is fine (well, when bufferReady is called at all), but then it gives a short buffer (like it's at the end of the file), and it just sits there.  Sometimes the decoding thread will crash, bringing down the whole app.
- WAV files don't seem to decode at all (MP3 files do, kind of).

One option that others on the internet have used is to just decode using FFMPEG, and not use QAudioDecoder at all.
Then, the rest of the plan still works:  use QAudioOutput (now called QAudioSink), and do writes (push-style) to the associate QAudioDevice.  We can pull from the QBuffer, do EQ, pitch/tempo, etc. and push to the device.  Have to implement start/stop/pause to control the player thread, but I was going to do that anyway.

Another option (that doesn't involve FFMPEG) might be to take the source code for QMediaPlayer, and put in a hook for output processing, something like DSP_process(QBuffer in, uint32 in_offset, uint32 in_length, QBuffer out), and run all the EQ/etc in real time at playback time.  Then, the whole (working!) QMediaPlayer infrastructure would work to our benefit.  But this does mean recompiling at least QMediaPlayer.   Could this be subclassed and wrapped, such that the DSP_process is there as a hook function?  I don't know yet.

**Comment by mpogue2** [comment]
OK, in the Qt6 source tree, there is an example of a command line QAudioDecoder.  It's NOT in the normal list of examples that ships with Qt6.2 .

I compiled it, and it worked fine.  So, it's something that I am doing wrong in SquareDesk that is causing it to fail.

This example is here in the Qt6 tree:
/Users/mpogue/Qt6Source/qt5/qtmultimedia/examples/multimedia/audiodecoder

Output command line is here:
/Users/mpogue/Qt6Source/qt5/qtmultimedia/examples/multimedia/build-audiodecoder-Qt_6_2_1_for_macOS-Debug/audiodecoder.app/Contents/MacOS

NOTE:  It says "Qt5", but it's really Qt6.2 .  This is a known misnomer, apparently.

**Comment by mpogue2** [comment]
I pulled the AudioDecoder class into SquareDesk, instantiated it statically inside flexible_audio.c, and started it up from the flexible_audio constructor, just as a test.  Works fine.

**Comment by mpogue2** [comment]
OK, got rid of the WAV file output, and replaced it with a QBuffer/QByteArray implementation.  
Tell the decoder to setSource("filename.mp3"), and start(), and it will do its thing.  A couple of minute MP3 file takes up about 27MB (yes, the whole thing is read into memory, just once), and takes about 250ms to load.  Not bad!

Next step is to build a controllable Player (play/pause/resume/stop) in a separate thread that reads the QByteArray, and sends it to a QAudioSink (used to be called QAudioOutput).  Once that is working, we can add EQ, Force Mono, etc..

There are several ways to build this, and I think they are all tradeoffs between memory and CPU.  For example, if Force Mono is on, we could set the preferredFormat to MONO, and save all that memory.  Also, the EQ and BPM would go a lot faster, most likely.  BUT, if the Force Mono checkbox is changed, we'll then have to reload the whole song (including allocating new memory).  For now, I'm going to implement "read once, process in realtime at play time", and optimize memory later.  Force Mono, in particular, doesn't change that often, so that might be a nice optimization later.  In the short term, it should not take long to process Stereo -> Mono, and I'm already doing Volume processing in realtime anyway, so the number of additional cycles isn't that much.

We could also use QMediaPlayer (which I already have working) on a separate "processed" buffer (and whenever the pitch/tempo/EQ sliders change, we process the main buffer to the processed buffer), but that doubles the amount of memory.  I'm going to try to make my own Player to start, which is a bit more work, and uses more realtime CPU, but uses less memory, and experiences less latency when sliders change (don't have to process the whole file).  If we want to try to drop battery consumption during playback, and are OK with the additional memory, that's certainly another way to do it.

In the short term, I want to see if we do all the processing in real time, like the current player.

**Comment by mpogue2** [comment]
a0f40220c816b82331c258679a40e28e27a5d77e has a Proof of Concept for the QAudioDecoder code.  It seems to work fine on M1 Mac.

**Comment by mpogue2** [comment]
ed075fa2ed8630faa04bbef9047af49b1adf316e has a thread-based player that works pretty good.  It even stops at the end of the song.  

Unlike with QMediaPlayer, we have access to the data on the way from the PCM buffer to the speakers.  So, we can add realtime DSP-style processing for Volume sliders, Force Mono menu option, EQ sliders, pitch/tempo sliders, etc.

**Comment by mpogue2** [comment]
2435b691c9c0d1e64ad5d9706878c7e9487d27c7 implements Force Mono (on all the time right now), Volume slider, and Mix/Pan slider.
1bd6338ba2ddaaf7dc34c8ed3e7908b3146fcb98 implements song positioning, so that the user can see the songSlider, can move the handle, can click in it to jump the play head, etc.

There's a lot of experimental code commented out in there.  I'll clean it up shortly.
There are THREE implementations of a player in there right now:  QMediaPlayer (which can't get access to the PCM data), a player based on using a Timer to push into the QAudioSink (which skips a lot, but has access to PCM), and the latest hotness: the PlayerThread, which plays and DSP-processes the data.  Only the third one will survive after cleanup.

**Comment by mpogue2** [comment]
Note to self: I think kfr requires deinterlaced data for stereo processing, which is unfortunate.  It has a deinterlacer and an interlacer, but it just takes extra CPU cycles to do so, that I was hoping to avoid.

**Comment by danlyke** [comment]
Totally ridiculous observation, but.... I just noticed that SuperCollider builds with Qt. It likely does everything we want to do with audio, and everything we will ever want to do with audio and is embeddable into an application... Other than that moment of "aha" and "oh, whoah" I haven't looked at it further, but it is definitely on my list and I got the book at home to start learning how to use it.

**Comment by mpogue2** [comment]
52ce358ecb6e8071b6b3257257da58b58be24e18 implements BPM detection, and updating the Tempo slider accordingly.
As of 52535ebb5bc320a9967108fd5b3404b8bf07a2f1, EQ now works (should be nearly identical to the existing libbass EQ, and the B/M/T sliders are hooked up.  They work nicely in real time (without crashing the player thread!).

NOTE: both KFR (for EQ) and MiniBPM both required deinterleaved mono data.  Right now, everything is hard-coded to be mono output (that is, Force Mono setting is ignored, and everything is mono).  

Overall, the MP3 decoder (AudioDecoder) reads the entire MP3 file into memory, as 8-byte-per-frame = 2 channels of float format.  BPM just grabs from 10-20s of that audio, mixes it down to a temp buffer, and detects BPM from that.  Then, the mixdown (using the Mix slider), volume, and EQ (B/M/T + intelligibility) are all done at PLAYBACK time, using an 8192-sample mono float buffer, that is used for both KFR EQ and (soon) for RubberBand.

Just before samples are handed off to Mac OS for playback, they are re-interleaved to 2 stereo channels that contain identical data (so it sounds like mono, even though the OS wants the data in stereo format).

**Comment by mpogue2** [comment]
@danlyke SuperCollider is cool!  But, I don't think we have a use case for the kind of synthesis-from-code stuff that it does.  kfr (EQ, Intelligibility Boost) and miniBPM (BPM detection) seem to be well optimized, and are open source.  They seem to work.  The last piece we need on the audio side is independent pitch/tempo processing, and I hope to try RubberBand shortly.  The RubberBand demo shows that it's pretty mature (and open source), so hopefully it will work OK, too.

Did you have some specific use cases in mind for SuperCollider?

**Comment by mpogue2** [comment]
219ee4bcfd633ad3a83256c9ba21c039e9dc111a implements RubberBand, but the sample passing around is not quite right yet.  The pitch slider is hooked up and works, but the output is not smooth yet (even with mode::smooth), and it IS smooth in the RubberBand Audio Demo app, so it's something that I am doing wrong.   The tempo slider is not hooked up yet, because it shows the sample-passing-around problem even more.

So, it needs work, but it's getting closer.

Still TODO:
```
// TODO: PITCH/TEMPO (rubberband) ********
// TODO: VU METER (kfr) ********
// TODO: LOOPS/sample accurate loops (myPlayer)
// TODO: SOUND EFFECTS (later) ********
// TODO: allow changing the output device (new feature!) *****

```

**Comment by danlyke** [comment]
I don't yet have specific use cases for SuperCollider, but my read was that
it can read MP3 files and do all of the transforms we're talking about
(pitching, tempo, EQ, etc) and then a whole bunch more stuff as well.

And it's an open source project with a large user base, that's used by a
lot of other projects (I'm currently playing with Sonic Pi), so it's gonna
continue to be a thing for a while.

**Comment by danlyke** [comment]
Oh, wait, my bad: It looks like SuperCollider doesn't actually do the file
handling, that's something SonicPi or various other add-ons are layering on
top of it.

**Comment by danlyke** [comment]
Or, well, now I'm not sure what to believe any more. I was skimming through the SuperCollider book when I wrote that previous comment, but...: https://doc.sccode.org/Classes/SoundFile.html

**Comment by mpogue2** [comment]
Turns out that BreakfastQuay (RubberBand folks) have a sample app, with source, that basically does exactly what I was writing with the AudioDecoder and PlayerThread class.  Except theirs (with a few modifications) works fine, and mine has problems.

The demo depends on a couple of other small open source libraries (well, they're more like blobs of code).  But, these blobs claim to be portable to Mac, Windows, and Linux.  They seem to work fine on Mac.  I was able to compile everything (using CMake and make, not with their Qt .pro file), and it works fine.

Modifications:
The demo had only tempo control, and I added pitch control as well.
The demo used standard RubberBand defaults, and I wanted Smooth, so that a sine wave comes out nice and clean.

Because this sample class is almost identical in function to what I just spent many hours writing, I am tempted to throw away mine, and just adopt this one.  If I can get it into .pro format for building, I suspect it will work 100%.  The other advantage is that the demo code implement full multichannel, whereas I was implementing just Force Mono for now.

What the demo does not do is the EQ, but I think I can just drop that code in from my version (since it seems to work).  I can also implement Mix and Volume control using the same approach, I think.

**Comment by mpogue2** [comment]
One thing I notice is that once I go away from Tempo = 1.0, Pitch = 1.0, the stereo image is spread way wide.  It doesn't go back to normal either, when the Tempo/Pitch are put back, until Pause/Resume is done.

The audio quality from RubberBand is definitely lower than lib bass's pitch/tempo algorithm.  The phasing seems to be caused by choosing "Smooth" as the crispness.  But, there is no crispness setting that seems to make it sound as good as libbass.  I missed this in the original evaluation of RubberBand.

**Comment by mpogue2** [comment]
[Example.app.zip](https://github.com/mpogue2/SquareDesk/files/7574130/Example.app.zip)

Here is the Sample App, with the modifications I made.  I suspect it will just run on an M1-based Mac OS X.

**Comment by mpogue2** [comment]
Alternatives?
https://www.surina.net/soundtouch/soundstretch.html#examples
superPowered (web browser version has similar problems to RubberBand).

AH HAH!
https://documentation.help/BASS_FX/TEMPO-About%20Algorithms.htm
States that libbass_fx uses SoundTouch.  So, I will check that out.  If I can achieve as good as libbass_fx, I think we're done.

**Comment by mpogue2** [comment]
Yeah, SoundTouch looks like the right one for us.  For small values of pitch and tempo change (within the range that we need for square dancing), it sounds GREAT -- it sounds identical to what we have now.  

It was an easy compile (cmake . ; make -j).  Generated a static library file, libSoundTouch.a, and an executable to test (soundstretch).  The whole thing compiled in LITERALY 3.4 SECONDS on the M1.  

WOW.

**Comment by mpogue2** [comment]
SoundTouch also has BPM detection (essentially what we have now).  I don't know how it compares to MiniBPM, which also seems to work.

**Comment by mpogue2** [comment]
As of 61a3ac1fe5b43c58b3f0062b486cb3bb0425092f, SoundTouch is basically working, modulo bugs.  It's definitely better than RubberBand.  EQ turned off right now (needs deinterleaving, or redesign inside kfr).

**Comment by mpogue2** [comment]
As of e4e08fb8a0ba4a4fdf54a84d19d79e83b01f5ca3, EQ is now working, with both Stereo and Mono paths.  Each path is optimized to do only what it needs.

**Open/known bugs:**
1. Still a periodic loss of samples ("glitches") when tempo is higher than base tempo.
2. I haven't implemented flush() at the very end of the song.
3. M1 on M1 CPU utilization is up to about 25% (stereo) and 21% (mono), compare to X86 on M1 = 13% (stereo/mono).
Compare to M1 on M1 without the 4 BiQuads: 18% (stereo/mono).  So much of the difference (3 to 6%, which is half the difference) is in the biquad implementation.  Might want to optimize the biquads a bit, e.g. if B/M/T/intel are set to 0dB -- just remove them entirely from the bq list.  For X86 on M1, the EQ sliders have no effect on CPU utilization.
4. When the audio is slowed down, the delay between slider change and audible effect of that slider goes UP.  This is true for EQ sliders, Mix slider, and Volume slider.  This is a good argument for putting those AFTER the soundTouch pitch/tempo processing.  But that means there will still be a delay for pitch/tempo changes that can be annoying.

**Comment by mpogue2** [comment]
5. I have not tested this code on X86, only on Native M1 on M1.

**Comment by mpogue2** [comment]
As of ecfbc08d6b5b58e0e0befb74df8356fbef8f6e27, basic loop functionality is working.  Within a session I can set intro and outro points, and the looping works correctly.  I also reworked loadMP3file() a bit, to handle the now asynchronous loading of the MP3 file.

BUG: intro/outro points, and the loop on/off settings are not persistent, although pitch is properly persistent.

**Comment by mpogue2** [comment]
Similar (maybe related) to Bug #1 above:  if tempo << 100%, then slider changes (Mix, Volume, EQ) take a LONG time to take effect.  In fact, they take longer and longer over time.

I can move the Mix/Volume/EQ to be AFTER the SoundTouch stage, but that only fixes those slider.  The Pitch/Tempo sliders will still take a while for changes to propagate, because presumably samples are backing up inside SoundTouch.  they just are not being consumed fast enough by the AudioSink.  Not sure how to fix this, but it's related, I think, to the fact that the AudioSink runs out of samples (I think) when the tempo slider is >>100%.  I'm doing this realtime SoundTouch thing wrong.

**Comment by mpogue2** [comment]
Yep, I was doing the SoundTouch thing wrong.
6e6de06b2918aa7018178dfb69bf8f3c75513a21 fixes both tempo < 100% lag and tempo > 100% audio glitching.
**We basically have a complete music player now, with pitch/tempo/bass/mid/treble/volume/mix all working properly.**

The key idea was that we need to scale the number of samples going INTO the SoundTouch stretcher/squisher, so that we get the correct number of samples coming OUT (the number needed by the AudioSink right now).  

A second key idea was that we need to scale the number of samples BEFORE the EQ stage, because the EQ filters also have state, and we don't want to process N samples, and use only M < N of them, and then resubmit the same N-M samples to the filters a second time. That messes up the filter, which wants a continuous stream of samples with no gaps and no repeats.  So, scaling the number of samples used down/up from the number requested by the AudioSink fixed all the problems.  There's a handy SoundTouch function called getInputOutputSampleRatio() that gives us exactly the number we need for this.

**Comment by mpogue2** [comment]
VU meter now works, as of e5747ab0c04b6efbd11fcebec791925a10a98760 .

**Comment by mpogue2** [comment]
### Current TODO lists:

**Bugs:**
1. EQ sliders are scratchy (limit filter changes to when we're not using them)

**Not implemented yet:**
1. Sound FX
2. Fade
4. Compressor
5. Intelligibility Boost

**Cleanup:**
1. Combine AudioDecoder and flexible_audio classes.  No need for both.
2. Rename playPosition_samples -> _frames
3. Reduce CPU utilization (perhaps with ARM NEON for the biquads?)

**New features:**
1. Allow output device to change.  Or, at MINIMUM, allow it to change while SquareDesk is running, e.g headphones plugged in, Sound preferences panel changes system sound, so that we don't need to restart it.  Would this also allow Zoom callers to not have to get Loopback?
2. Logarithmic volume?
3. Logarithmic VU meter?

**Comment by mpogue2** [comment]
Hard limiter added: 6648d0e0b4e4baa7c381de241b493d67d485d5c3 .  Sounds a bit better when EQ are turned up too high.  Still bad, just not as bad.

**Comment by mpogue2** [comment]
As of 1926f16b4d3e72e4fea626f4e7c35f5ab968d206, the entire music player seems to be working very reliably, including all the sliders (which operate in realtime, just like the X86 version), and the sound FX (most important one being the "long tip indicator", which I use all the time).

Still not working (TODO):
1. Reference pages (Qt bug is preventing them from coming up).
2. SD is complaining that the database is out of date ("Database format version (341) is not the required one (396)").  Probably just need to copy the right one up there.  I could swear this was working correctly before.
3. Compressor is not implemented at all right now.  kfr doesn't have one, so this is probably some heavy lifting.
4. Intelligibility Boost is fairly straightforward (except for the Q part, which I have to look up), but not implemented.
5. I do want to consolidate the flexible_audio class and the AudioDecoder class into a single flexible_audio class, that is essentially a drop-in replacement for the bass_audio class we use on X86.  (cleanup....)
6. Time to remove ReplayGain all the way.
7. I think that the taglib is not reading BPM yet for BPM override.  Not sure -- I'll have to check.  taglib is currently flagging an error on song "Awoo"s ID3v2 tags, surprisingly.
8. The build might not be copying all the files in, like it normally does.  Gotta check on that.  I had to copy some stuff in manually (maybe I did that only for the Debug build??).

It's definitely usable as a music player now, although not all features are up on M1 Native yet.  Sounds good, and seems very reliable at this point.  It's nice to see the UI bugs gone, as per Qt6.

I'd like to get the above problems fixed, and then backport the whole thing to X86, replacing what we have now.  That would accomplish 3 goals:

1. getting rid of libbass, which is Closed Source
2. moving to Qt6 (which fixes a bunch of the annoying UI bugs in Qt5, which isn't doing bug fixes anymore!)
3. running M1 Native on the M1 Macs.

I'll try not to merge the branch back into mainline until I have most of the features implemented, so that we don't go backwards in features (if possible).

And by the way, it's SO nice to use the M1 laptop as a build machine.  It's lightning fast at compiling.  With MenuMeters I can watch all 8 CPU's madly compiling in parallel.  Speed is very, very impressive.  I wonder if @danlyke could compile this on Linux in a VM on a Mac?

**Comment by mpogue2** [comment]
Yep, the SD problem in the Release build was an old sd_calls.dat file.  Works fine now with a manual copy.  This has got to be a build problem.

**Comment by mpogue2** [comment]
SoundTouch is still 100% stable on the M1 build (FlexibleAudio branch) as of 71b81e94182df9571d1c07a65b997c9eb7a900a7 .

**Comment by mpogue2** [comment]
OK, everything is in, and flexibleAudio was merged into Master.  Everything seems very stable.
So, closing as FIXED.
